src/cascadia/TerminalConnection/AzureConnection.cpp

```diff
@@ -730,14 +746,38 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
         _WriteStringWithNewline(RS_(L"AzureSuccess"));
 
         // Request for a terminal for said cloud shell
-        const auto shellType = _ParsePreferredShellType(settingsResponse);
+        auto shellType = _ParsePreferredShellType(settingsResponse);
         _WriteStringWithNewline(RS_(L"AzureRequestingTerminal"));
-        const auto socketUri = _GetTerminal(shellType.value_or(L"pwsh"));
+        const auto socketUri = _GetTerminal(shellType);
         _TerminalOutputHandlers(L"\r\n");
 
-        // Step 8: connecting to said terminal
-        const auto connReqTask = _cloudShellSocket.connect(socketUri);
-        connReqTask.wait();
+        //// Step 8: connecting to said terminal
+        {
+            wil::unique_winhttp_hinternet sessionHandle, connectionHandle, requestHandle, socketHandle;
+            Uri parsedUri{ socketUri };
+            sessionHandle.reset(WinHttpOpen(HttpUserAgent.data(), WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY, nullptr, nullptr, 0));
+            THROW_LAST_ERROR_IF(!sessionHandle);
+
+            connectionHandle.reset(WinHttpConnect(sessionHandle.get(), parsedUri.Host().c_str(), INTERNET_DEFAULT_HTTPS_PORT, 0));
+            THROW_LAST_ERROR_IF(!connectionHandle);
+
+            requestHandle.reset(WinHttpOpenRequest(connectionHandle.get(), L"GET", parsedUri.Path().c_str(), nullptr, nullptr, nullptr, WINHTTP_FLAG_SECURE));
+            THROW_LAST_ERROR_IF(!requestHandle);
+
+            THROW_IF_WIN32_BOOL_FALSE(WinHttpSetOption(requestHandle.get(), WINHTTP_OPTION_UPGRADE_TO_WEB_SOCKET, nullptr, 0));
+#pragma warning(suppress : 26477) // WINHTTP_NO_ADDITIONAL_HEADERS expands to NULL rather than nullptr (who would have thought?)
+            THROW_IF_WIN32_BOOL_FALSE(WinHttpSendRequest(requestHandle.get(), WINHTTP_NO_ADDITIONAL_HEADERS, 0, nullptr, 0, 0, 0));
+            THROW_IF_WIN32_BOOL_FALSE(WinHttpReceiveResponse(requestHandle.get(), nullptr));
+
+            socketHandle.reset(WinHttpWebSocketCompleteUpgrade(requestHandle.get(), 0));
+            THROW_LAST_ERROR_IF(!socketHandle);
+
+            requestHandle.reset(); // We no longer need the request once we've upgraded it.
+            // We have to keep the socket session and connection handles.
+            _socketSessionHandle = std::move(sessionHandle);
+            _socketConnectionHandle = std::move(connectionHandle);

```

Comment:
Why do we need to keep `_socketSessionHandle` and `_socketConnectionHandle`?

Reply:
It's part of the winhttp contract. :)